### ISSUE TITLE
Conventions for options, in particular those affecting binary compatibility

### ISSUE THREAD

**Issue by sigiesec** [bincrafters/community#491]
I am curious if there is a standard set of options that should be provided by all packages, in particular those affecting binary compatibility.

For example, several packages have options "shared" or "fPIC". For the latter, a dedicated documentation page exists (https://bincrafters.readthedocs.io/en/latest/howtos/add_fpic_option_to_recipe.html), which has the interesting introduction:

> At a certain point, we realized we should be adding fPIC option for many packages which we did not consider before. So, please check with the team in slack if you’re not sure whether or not you need fPIC.

I think this is somewhat too soft. Except for header-only libraries, I cannot imagine why this shouldn't be relevant.

It is also not clearly said the option *must* be named fPIC (and neither does the conan documentation, even though the cmake generator includes a special handling of it).

In addition, this is the only option that is mentioned at all in the documentation.

For Windows/Visual Studio resp. Visual C++ users, one critical aspect often is the provision of static libraries built for LTCG (link-time code generation). I noticed that many bincrafters packages lack any options to build libraries for LTCG. [Just as a side note, since I am not an expert for gcc/clang: While gcc/clang also have this feature, AFAIK it is less critical for producing efficient code, so maybe it is fine to neglect it for gcc/clang.]

Maybe there are several aspects here:
* Are there conventions that define which name an option should have *if* it is implemented by some package? (e.g. that the correct form is "fPIC", not "fPic" or "positionIndependent" or whatever) It is clear that there are custom options that are specific to some package, but there are several options (probably many more than I mentioned) that can be defined in a general way. I think this is essential for gaining acceptance, since users don't want to look into the details of each package to figure this out for each individual library, and contributors to package definitions don't need to think about this individually.
* Are there conventions that define which options should be provided by a package? Maybe there could be a few levels ("basic", "sophisticated") that can be distinguished. For example, it could be required that all "basic" packages offer "shared" and "fPIC" options already, and "sophisticated" packages offer "static-ltcg" in addition. I think this is also essential for gaining acceptance, since users don't want to check this for each individual library, if they consume a number of them.

**Comment by Croydon** [comment]
I agree that these common options, which often have specific handling in generators and tools, should get better documentation.

**Comment by uilianries** [comment]
The `fPIC` convention was defined long time ago with the Conan team. At the beginning, we started to use `with_fpic`, but at some point it was replaced by fPIC.

We try to keep fPIC because both Conan helpers for CMake and autotools use that option:

https://github.com/conan-io/conan/blob/84dc7a0b7dd9e2c26a2065829323af4246370ea3/conans/client/build/cmake_flags.py#L300

In addition, the Conan Community team follows the fPIC and we try to stay in line with them.

About our documentation it could be outdated, Conan is growing fast with new features.
